### PR TITLE
feat: update MeshAccessLog API to align with the memo

### PR DESCRIPTION
**Pull request by lobkovilya**
* Change MeshAccessLog policy to align with [memo](https://github.com/kumahq/kuma/blob/master/docs/guides/new-policies.md#how-to-map-api-to-a-go-struct)
* Convert schema validation error to Kuma ValidationError
* Add `SafeDeref` method for easy access to fields that we don't need to distinguish nil/zero values
* Add unit tests for MeshAccessLog validation and configurer

Signed-off-by: Ilya Lobkov <ilya.lobkov@konghq.com>

### Checklist prior to review

- [X] Link to docs PR or issue --
- [X] Link to UI issue or PR --
- [X] Is the [issue worked on linked][1]? --
- [X] The PR does not hardcode values that might break projects that depend on kuma (e.g. "kumahq" as a image registry) --
- [X] The PR will work for both Linux and Windows, system specific functions like `syscall.Mkfifo` have equivalent implementation on the other OS --
- [X] Unit Tests --
- [X] E2E Tests --
- [X] Manual Universal Tests --
- [X] Manual Kubernetes Tests --
- [X] Do you need to update [`UPGRADE.md`](../blob/master/UPGRADE.md)? --
- [X] Does it need to be backported according to the [backporting policy](../blob/master/CONTRIBUTING.md#backporting)? --
- [X] Do you need to explicitly set a [`> Changelog:` entry here](../blob/master/CONTRIBUTING.md#submitting-a-patch) or add a `ci/` label to run fewer/more tests?

[1]: https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword

### CODE REVIEW

File: pkg/core/resources/model/utils.go [hunk 1]
```diff
@@ -96,3 +96,11 @@ func IsEmpty(spec ResourceSpec) bool {
...
 	return &t
 }
+
+func SafeDeref[T any](ptr *T) T {
+	if ptr == nil {
+		var zero T
+		return zero
+	}
+	return *ptr
+}
```
Comment (jakubdyszkiewicz) [inline pkg/core/resources/model/utils.go:106]: I can't see how this is connected with our core model. I'd put it into `pkg/utils/pointers/pointers.go` and then use `pointers.To(X)` and `pointers.SafeDeref(*X)` or even `pointers.Deref` to be shorter.

File: pkg/plugins/policies/meshaccesslog/plugin/v1alpha1/plugin_test.go [hunk 2]
```diff
@@ -25,6 +26,11 @@ import (
 )
 
 var _ = Describe("MeshAccessLog", func() {
+
+	core.TempDir = func() string {
+		return "/tmp"
+	}
```
Comment (lahabana) [inline pkg/plugins/policies/meshaccesslog/plugin/v1alpha1/plugin_test.go:32]: Don't you need to restore the TempDir?

File: pkg/core/resources/model/utils.go [hunk 3]
```diff
@@ -96,3 +96,11 @@ func IsEmpty(spec ResourceSpec) bool {
 func PtrTo[T any](t T) *T {
 	return &t
 }
+
+func SafeDeref[T any](ptr *T) T {
```
Comment (michaelbeaumont) [inline pkg/core/resources/model/utils.go:100]: Can we try to have this stuff in separate PRs? This has nothing to do with `MeshAccessLog`

Comment (lobkovilya) [reply pkg/core/resources/model/utils.go:100]: Probably it can be a separate PR, but it should go before this because this func is actively used

Comment (lobkovilya) [reply pkg/core/resources/model/utils.go:100]: https://github.com/kumahq/kuma/pull/5586

File: pkg/core/resources/model/utils.go (hunk 1 above)
Comment (lobkovilya) [reply pkg/core/resources/model/utils.go:106]: Moved to another PR https://github.com/kumahq/kuma/pull/5586

File: pkg/plugins/policies/meshaccesslog/plugin/v1alpha1/plugin_test.go (hunk 2 above)
Comment (lobkovilya) [reply pkg/plugins/policies/meshaccesslog/plugin/v1alpha1/plugin_test.go:32]: Each spec runs as a separate process, so we don't have to restore the TempDir, but this code should be inside `BeforeEach`:
```go
func BeforeEach() {
	core.TempDir = func() string {
		return "/tmp"
	}
}
```
I'll fix this in other places as well

Comment (lobkovilya) [reply pkg/plugins/policies/meshaccesslog/plugin/v1alpha1/plugin_test.go]: Fixed

File: pkg/core/resources/model/rest/unmarshaller.go [hunk 4]
```diff
@@ -84,25 +81,6 @@ func (u *unmarshaler) Unmarshal(bytes []byte) (Resource, error) {
 	return restResource, nil
 }
 
-func (u *unmarshaler) rawSchemaValidation(bytes []byte, schema *spec.Schema) error {
```
Comment (michaelbeaumont) [inline pkg/core/resources/model/rest/unmarshaller.go:87]: Also unrelated to MeshAccessLog isn't it? No reason to block the PR but really should be in its own PR for visibility/understanding/traceability

**Review by michaelbeaumont — APPROVED** [review]

File: pkg/plugins/policies/meshaccesslog/api/v1alpha1/meshaccesslog.go [hunk 5]
```diff
@@ -51,24 +49,24 @@ type Backend struct {
...
-	Path string `json:"path,omitempty"`
+	Path string `json:"path"`
 }
 
 type Format struct {
-	Plain           string      `json:"plain,omitempty"`
-	Json            []JsonValue `json:"json,omitempty"`
-	OmitEmptyValues bool        `json:"omitEmptyValues,omitempty"`
+	Plain           *string      `json:"plain,omitempty"`
+	Json            *[]JsonValue `json:"json,omitempty"`
```
Comment (jakubdyszkiewicz) [inline pkg/plugins/policies/meshaccesslog/api/v1alpha1/meshaccesslog.go:68]: slice can already be nil, do we need a pointer here?

File: pkg/core/resources/model/rest/unmarshaller.go (hunk 4 above)
Comment (lobkovilya) [reply pkg/core/resources/model/rest/unmarshaller.go:87]: Right, it was a small problem with validation that I noticed after the changes in MeshAcessLog. Should've opened another PR

File: pkg/plugins/policies/meshaccesslog/api/v1alpha1/meshaccesslog.go (hunk 5 above)
Comment (jakubdyszkiewicz) [reply pkg/plugins/policies/meshaccesslog/api/v1alpha1/meshaccesslog.go:68]: https://go.dev/play/p/5gAykZVuEAS

Comment (michaelbeaumont) [reply pkg/plugins/policies/meshaccesslog/api/v1alpha1/meshaccesslog.go:68]: Finally somebody asks the big question about slices! 

I suspect that the generation code as is expects a pointer type.

IMO we should use a pointer if we can, as long as we really treat `xs := nil` just like `xs := []X{}` everywhere. I think we just have to be careful/make sure that we never serialize an above value to `{ json: null }`, for example what happens with:

```
var values []JsonValue
f := Format{ Json: &values }
```


? I'm not positive.

Comment (michaelbeaumont) [reply pkg/plugins/policies/meshaccesslog/api/v1alpha1/meshaccesslog.go:68]: https://go.dev/play/p/Nfpf30MCi0J indicates that this is dangerous

Comment (jakubdyszkiewicz) [reply pkg/plugins/policies/meshaccesslog/api/v1alpha1/meshaccesslog.go:68]: > IMO we should use a pointer if we can

But why, what is the advantage?
Because I can see a disadvantage of writing extra code that has to deal with nils. Also, it's easier to get panic on nil dereference, because Go has no mechanism to protect us from it (except default values if you drop pointers).

Comment (michaelbeaumont) [reply pkg/plugins/policies/meshaccesslog/api/v1alpha1/meshaccesslog.go:68]: The advantage in the abstract is of course that the semantics of a non-pointer slice would be the same as any non-pointer value. It's absolutely insane to create a type where there are 2 different "empty lists", aka values of length 0 but that's what go did.

> Because I can see a disadvantage of writing extra code that has to deal with nils

That's what I meant, if we can, given go as is. So if the danger of serializing a nil slice as `null` exists, we can't use a pointer wrapper.

> Also, it's easier to get panic on nil dereference, because Go has no mechanism to protect us from it (except default values if you drop pointers).

Isn't the slice case the only one where we _can't_ get a panic because of a nil dereference? I.e. we have to deal with it for every other type.

Comment (lobkovilya) [reply pkg/plugins/policies/meshaccesslog/api/v1alpha1/meshaccesslog.go:68]: If you have:
```go
type Spec struct {
    MySlice []MyType `json:"mySlice,omitempty"`
}
```
then when you serialize to json it won't distinguish between nil and empty slice. Both will be serialized to:
```json
"spec": {}
```
There is a proposal to add `omitnil` https://github.com/golang/go/issues/22480, but it doesn't seem like it's going to happen

Comment (michaelbeaumont) [reply pkg/plugins/policies/meshaccesslog/api/v1alpha1/meshaccesslog.go:68]: That's bad, right? Because we need to have a value that means "not set". So the only way to express an optional slice, without the danger of empty slice vs nil slice, would be 

```
    MySlice []MyType `json:"mySlice"`
```

Comment (michaelbeaumont) [reply pkg/plugins/policies/meshaccesslog/api/v1alpha1/meshaccesslog.go:68]: Maybe there's a better JSON encoding lib for go

Comment (lobkovilya) [reply pkg/plugins/policies/meshaccesslog/api/v1alpha1/meshaccesslog.go:68]: Yes, but you also need annotations:
```
// +optional
// +nullable
 MySlice []MyType `json:"mySlice"`
 ```
 and in that case, it will always serialize omitted by user values to `mySlice: null`

Comment (lobkovilya) [reply pkg/plugins/policies/meshaccesslog/api/v1alpha1/meshaccesslog.go:68]: That's actually what we had in MeshAccessLog before this PR https://github.com/kumahq/kuma/pull/5580/files#diff-ed8e7e0fddb480be4d601ae35bb953cddd53659fb3b9d306b43996a8f322cb38L40

Comment (michaelbeaumont) [reply pkg/plugins/policies/meshaccesslog/api/v1alpha1/meshaccesslog.go:68]: So I don't care about which annotations we have to use but what I want ideally is

```
type Struct struct {
  MySlice *[]MyType
}
```

where a `nil` pointer means not set, just like for every other possible field type. Where a nil slice and an empty slice are the exact same thing. So I can just use `val.MySlice == nil` to check whether the value is provided and `len(val.MySlice) == 0` to check whether an empty slice has been set.

and where encoded values are either `{}`, `{mySlice: []}`, or `{mySlice: [...]}` and there is no `{mySlice: null}`

Comment (lobkovilya) [reply pkg/plugins/policies/meshaccesslog/api/v1alpha1/meshaccesslog.go:68]: > I don't know if it'd violate our generated open API specs

Yes, If I recall correctly it violates the specs without `// +nullable`

Comment (jakubdyszkiewicz) [reply pkg/plugins/policies/meshaccesslog/api/v1alpha1/meshaccesslog.go:68]: Ok

**Review by jakubdyszkiewicz — APPROVED** [review]
